src/lib/allocation_tracker.cc

```diff
@@ -0,0 +1,365 @@
+  uint64_t regs[PERF_REGS_COUNT];
+  /* if PERF_SAMPLE_REGS_USER */
+  uint64_t size;                          /* if PERF_SAMPLE_STACK_USER */
+  std::byte data[PERF_SAMPLE_STACK_SIZE]; /* if PERF_SAMPLE_STACK_USER */
+  uint64_t dyn_size;                      /* if PERF_SAMPLE_STACK_USER &&
+                                        size != 0 */
+};
+
+struct LostEvent {
+  perf_event_header hdr;
+  uint64_t id;
+  uint64_t lost;
+};
+
+class ReentryGuard {
+public:
+  ReentryGuard() : _ok(!_reentry_guard) { _reentry_guard = true; }
+  ~ReentryGuard() {
+    if (_ok) {
+      _reentry_guard = false;
+    }
+  }
+
+  explicit operator bool() const { return _ok; }
+
+private:
+  static thread_local bool _reentry_guard;
+  bool _ok;
+};
+
+thread_local bool ReentryGuard::_reentry_guard;
+
+static AllocationTracker *s_instance;
+static std::atomic<bool> s_track_allocations{false};
+static std::atomic<bool> s_track_deallocations{false};
+
+// Per-thread counters
+static thread_local int64_t s_remaining_bytes{0};
+static thread_local bool s_remaining_bytes_initialized{false};
+static thread_local pid_t s_tid;
+
+static pid_t s_pid;
+
+// count number of lost events
+static uint64_t s_lost_count;
+static std::mutex s_mutex;
+
+// Temporary struct used to write event data before pushing it on ring buffer
+// Cannot be on the stack otherwise it messes up stack traces:
+// it contains a buffer to store the stacktrace, and thus would use by itself
+// all the space allocated for the capturing the stacktrace
+static AllocationEvent s_event;
+
+static void allocation_tracking_free_locked();
+
+void track_allocation(void *addr, size_t size) {
+  s_remaining_bytes += size;
+  if (likely(s_remaining_bytes < 0)) {
+    return;
+  }
+
+  if (likely(s_track_allocations.load(std::memory_order_relaxed))) {
+    s_instance->track_allocation(addr, size);
+  } else {
+    // allocation tracking is disabled, reset state
+    s_remaining_bytes_initialized = false;
+    s_remaining_bytes = 0;
+  }
+}
+
+void track_deallocation(void *addr) {}
+
+class RingBufferWriter {
+public:
+  explicit RingBufferWriter(RingBuffer &rb) : _rb(rb) {
+    _tail = __atomic_load_n(&_rb.region->data_tail, __ATOMIC_ACQUIRE);
+    _head = _rb.region->data_head;
+    assert(_tail <= _head);
+    _available_size = _rb.data_size - (_head - _tail);
+  }
+
+  ~RingBufferWriter() {
+    __atomic_store_n(&_rb.region->data_head, _head, __ATOMIC_RELEASE);
+  }
+
+  RingBufferWriter(const RingBufferWriter &) = delete;
+  RingBufferWriter &operator=(const RingBufferWriter &) = delete;
+
+  inline size_t available_size() const { return _available_size; }
+
+  void write(Buffer buf) {
+    assert(buf.size() <= _available_size);
+
+    uint64_t head_linear = _head & _rb.mask;
+    char *dest = const_cast<char *>(_rb.start) + head_linear;
+
+    // check if we need to wrap around end of buffer
+    // \fixme remove memcpy by writing directly into the ring buffer ?
```

Comment:
What would be the advantage of suppressing `memcpy()`?

Reply:
I see 2 advantages:
* do less work: directly write to ringbuffer
* remove the need for static variable `s_event`